### PR TITLE
CT-2236 Refactor case creation with intro of Offender SAR

### PR DESCRIPTION
**Pull request by mseedat-moj**
## Description
Significant refactor of the way the `cases_controller.rb` operates with a view to introducing `Case::Base::SAR::Offender` model and related controller. This work is in tandem with #785 . 

Includes fix for https://dsdmoj.atlassian.net/browse/CT-2319.

This PR is based on the work carried out during #727 .

## Self-review checklist

* [ ] (1) Quick stakeholder demo done
* [ ] (2) Bug with before and after screenshots
* [x] (3) Tests passing
* [x] (4) Branch ready to be merged (not work in progress)
* [ ] (5) No superfluous changes in diff
* [ ] (6) No TODO's without new ticket numbers
* [ ] (7) Table indexes added if required

## Notes
This PR represents the unravelling of `cases_controller` into a more resource-based API. 

This is the first significant refactor since the system was operational 2 years ago. As a result the changes are made to ensure the system operates as-is without any changes to the expected User Experience or behaviours.

Rubycritic shows the difference in refactoring has had on the `cases_controller.rb` file:

cases-controller-rubycritic

Running `rails stats` before and after shows the way complexity of controllers has decreased overall, whilst the code size has grown to accomodate the increase in the number of controllers:

**MASTER**
```
+----------------------+--------+--------+---------+---------+-----+-------+
| Name                 |  Lines |    LOC | Classes | Methods | M/C | LOC/M |
+----------------------+--------+--------+---------+---------+-----+-------+
| Controllers          |   2911 |   2439 |      15 |     229 |  15 |     8 |
| Helpers              |    394 |    346 |       0 |      37 |   0 |     7 |
| Jobs                 |     38 |     32 |       4 |       3 |   0 |     8 |
| Models               |   4477 |   2872 |      47 |     405 |   8 |     5 |
| Mailers              |    166 |    132 |       3 |      10 |   3 |    11 |
| Channels             |     12 |      8 |       2 |       0 |   0 |     0 |
| Javascripts          |    875 |    626 |       0 |      77 |   0 |     6 |
| Libraries            |   3054 |   2521 |      23 |     205 |   8 |    10 |
| Tasks                |    692 |    566 |       1 |      12 |  12 |    45 |
| Mailer specs         |    395 |    338 |       0 |       0 |   0 |     0 |
| Decorator specs      |    944 |    787 |       0 |       1 |   0 |   785 |
| Feature specs        |   6833 |   5298 |       0 |      37 |   0 |   141 |
| Model specs          |   8657 |   6608 |       0 |      25 |   0 |   262 |
| Policy specs         |   1372 |   1156 |       0 |       0 |   0 |     0 |
| State_machine specs  |   9946 |   8782 |       4 |      39 |   9 |   223 |
| Factory_spec specs   |   1543 |   1298 |       0 |       4 |   0 |   322 |
| Lib specs            |    567 |    473 |       1 |       4 |   4 |   116 |
| Validator specs      |    798 |    674 |       0 |       0 |   0 |     0 |
| Routing specs        |    171 |    137 |       0 |       0 |   0 |     0 |
| Job specs            |    104 |     81 |       0 |       0 |   0 |     0 |
| Controller specs     |   8757 |   7233 |       0 |       6 |   0 |  1203 |
| View specs           |   5050 |   3789 |       0 |      40 |   0 |    92 |
| Helper specs         |    554 |    446 |       0 |       0 |   0 |     0 |
| Service specs        |  10519 |   8884 |       3 |      35 |  11 |   251 |
+----------------------+--------+--------+---------+---------+-----+-------+
| Total                |  68829 |  55526 |     103 |    1169 |  11 |    45 |
+----------------------+--------+--------+---------+---------+-----+-------+
  Code LOC: 9542     Test LOC: 45984     Code to Test Ratio: 1:4.8
```

**REFACTOR (CT-2236)**
```
+----------------------+--------+--------+---------+---------+-----+-------+
| Name                 |  Lines |    LOC | Classes | Methods | M/C | LOC/M |
+----------------------+--------+--------+---------+---------+-----+-------+
| Controllers          |   3367 |   2755 |      32 |     281 |   8 |     7 |
| Helpers              |    403 |    349 |       0 |      38 |   0 |     7 |
| Jobs                 |     38 |     32 |       4 |       3 |   0 |     8 |
| Models               |   4617 |   2991 |      48 |     417 |   8 |     5 |
| Mailers              |    166 |    132 |       3 |      10 |   3 |    11 |
| Channels             |     12 |      8 |       2 |       0 |   0 |     0 |
| Javascripts          |    875 |    626 |       0 |      77 |   0 |     6 |
| Libraries            |   3060 |   2526 |      23 |     205 |   8 |    10 |
| Tasks                |    698 |    571 |       1 |      12 |  12 |    45 |
| Mailer specs         |    395 |    338 |       0 |       0 |   0 |     0 |
| Decorator specs      |    957 |    787 |       0 |       1 |   0 |   785 |
| Feature specs        |   6858 |   5314 |       0 |      37 |   0 |   141 |
| Model specs          |   8997 |   6744 |       0 |      25 |   0 |   267 |
| Policy specs         |   1372 |   1156 |       0 |       0 |   0 |     0 |
| State_machine specs  |   9946 |   8782 |       4 |      39 |   9 |   223 |
| Support specs        |   3378 |   2774 |      10 |     125 |  12 |    20 |
| Factory_spec specs   |   1548 |   1302 |       0 |       4 |   0 |   323 |
| Lib specs            |    567 |    473 |       1 |       4 |   4 |   116 |
| Validator specs      |    798 |    674 |       0 |       0 |   0 |     0 |
| Routing specs        |    171 |    137 |       0 |       0 |   0 |     0 |
| Job specs            |    104 |     81 |       0 |       0 |   0 |     0 |
| Controller specs     |   7422 |   6159 |       0 |       4 |   0 |  1537 |
| View specs           |   4789 |   3580 |       0 |      37 |   0 |    94 |
| Form_model specs     |    121 |     96 |       0 |       0 |   0 |     0 |
| Helper specs         |    563 |    452 |       0 |       0 |   0 |     0 |
| Service specs        |  10494 |   8788 |       3 |      35 |  11 |   249 |
+----------------------+--------+--------+---------+---------+-----+-------+
| Total                |  71716 |  57627 |     131 |    1354 |  10 |    40 |
+----------------------+--------+--------+---------+---------+-----+-------+
  Code LOC: 9990     Test LOC: 47637     Code to Test Ratio: 1:4.8

```

The routing is setup to use more meaningful routes for cases post-creation. For example `/cases/sars/23/edit` and /cases/23/responses/new/upload_responses`.

It is hoped future additions to the application consider a more resource-orientated approach if possible.

For example, a SAR extension previously had the path `/cases/23/extend_sar_deadline`. This has been refactored with the creation of a new controller (SarExtensionsController) and corresponding resource-orientated path `/cases/23/sar_extensions/new`.

The aim of this work was:

- To consider reducing the complexity of the cases_controller in light of Offender Sar Work
- To make on-boarding of new developers easier when implementing new features/updating features
- To follow best practices with regards to smaller classes/methods and Single Responsibilty Principle

The aim of this work was not:
- To reduce the code size overall
- To make the code faster
- To improve implementation, revise Services/Models, etc

### Issues
- The routing is not the best. For example, the `case create` path generates a path appended with `_index` due to the way both `resources` and `resource` are declared for various case types
- Thee is inconsistency in the way CorrespondenceType abbreviations are arbitrary strings tryingt o describe 'collections' of similar case types and for Correspondence Type retrieval. These arbitrary strings are different to the case model string produced by Rails (`.model_name`) and therefore caused some problems with naming of controllers, view folders, polymorphic_paths, etc (some inconsistency).

### Related JIRA tickets
https://dsdmoj.atlassian.net/browse/CT-2236
https://dsdmoj.atlassian.net/browse/CT-2319

### Deployment
Requires data migrations of offender sars correspondence type

### Manual testing instructions
Ensure all areas of functionality are working as currently expected

### CODE REVIEW

File: lib/cts/cases/cli.rb [hunk 1]
```diff
@@ -116,7 +116,7 @@ def check(*args)
     option :responder, aliases: :r, type: :string,
            desc: 'ID or name of responder to use for case assignments.'
     option :responding_team, aliases: :t, type: :string,
-           desc: 'ID or name of responding team to use for case assignments.'
+           desc: 'ID or name of respondinag team to use for case assignments.'
```
Comment (Polsonby) [inline lib/cts/cases/cli.rb]: oops typo :)

File: app/controllers/cases/offender_sar_controller.rb [hunk 2]
```diff
@@ -34,19 +32,18 @@ def create
...
+
+      if @case.valid? && @case.save
+        session[:offender_sar_state] = nil
+        redirect_to case_path(@case.case) and return
+      end
+
+      @case.session_persist_state(create_params)
+      get_next_step(@case)
+      redirect_to step_case_sar_offenders_path + "/#{@case.current_step}"
     end
```
Comment (Polsonby) [inline app/controllers/cases/offender_sar_controller.rb]: Nice :)

File: spec/routing/cases_routes_spec.rb [hunk 3]
```diff
@@ -13,86 +13,86 @@
 
     context 'manager user' do
       let(:user) { manager }
-      it { should redirect_to '/cases/open' }
+      xit { should redirect_to '/cases/open' }
```
Comment (Polsonby) [inline spec/routing/cases_routes_spec.rb]: I'm a bit uncomfortable marking these specs as pending without more explanation as to why and a JIRA ticket to address it later. When we say, "it is now irrelevant", is that a temporary thing? Can we delete these specs instead?

Comment (mseedat-moj) [reply spec/routing/cases_routes_spec.rb]: Good spot - resolved by re-enabling and updating the routing. Separated out the 'filters' controller routes into a separate file

Comment (mseedat-moj) [reply spec/routing/cases_routes_spec.rb]: Goo spot - resolved

File: app/controllers/cases/responses_controller.rb [hunk 4]
```diff
@@ -47,7 +47,7 @@ def new
       authorize @case, @settings[:policy]
 
       @s3_direct_post = S3Uploader.s3_direct_post_for_case(@case, 'responses')
-      @approval_action = @settings[:approval_action]
+      @approval_action = @settings[:approval_action] # @todo: Unused value?
```
[the remark this answers is not in the document]
Comment (mseedat-moj) [reply app/controllers/cases/responses_controller.rb]: Yes good idea

File: spec/features/user_journeys/overturned_sar/trigger_spec.rb [hunk 5]
```diff
@@ -28,7 +28,8 @@
     CaseClosure::MetadataSeeder.unseed!
   end
 
-  scenario 'end-to-end journey', js: true do
+  # @todo: Test disable as is overturned FOI - requires revision
+  xscenario 'end-to-end journey', js: true do
```
[the remark this answers is not in the document]
Comment (mseedat-moj) [reply spec/features/user_journeys/overturned_sar/trigger_spec.rb]: Yes as discussed in slack

File: app/views/cases/select_type.html.slim [hunk 6]
```diff
@@ -13,5 +13,5 @@ ul#create-correspondences
 
   - @permitted_correspondence_types.each do |correspondence_type|
     li
-      = link_to correspondence_type.abbreviation_and_name,
-                new_polymorphic_url(correspondence_type.sub_classes.first)
+      - case_type = correspondence_type.abbreviation == 'ICO' ? 'case_ico' : correspondence_type.sub_classes.first
+      = link_to correspondence_type.abbreviation_and_name, new_polymorphic_url(case_type)
```
[the remark this answers is not in the document]
Comment (mseedat-moj) [reply app/views/cases/select_type.html.slim]: Yes this gave me major headaches and results in fragile code. Will setup a ticket

**Review by Polsonby — APPROVED** [review]

**Pull request merged**